[keyboard] bastardkb: restructure folder hierarchy

0xcharly opened the pull request:
Upstream changes to Bastard Keyboard boards.

This PR:

- Restructure folder hierarchy ahead of supporting other adapters/mcus
- Add support for 2 new keyboards:
  - Charybdis Mini (aka. Cmini, aka. Charybdis 3x6)
  - Dilemma
- Add support for adapter v2
- Add support for RP2040-based Splinky MCU

This PR is on behalf of @bstiq.

<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

Bastard Keyboards support for the following (adapter, mcu) pairs:

- [`v2/elitec`](https://github.com/Bastardkb/bastardkb-qmk/commit/40c5d056a3b3fac64d57ded3282701de8cf19509)
- [`v2/stemcell`](https://github.com/Bastardkb/bastardkb-qmk/tree/bkb-stemcell)
- [`v2/splinky`](https://github.com/Bastardkb/bastardkb-qmk/tree/bkb-rp2040)
- [`blackpill`](https://github.com/Bastardkb/bastardkb-qmk/commit/965b881baee82b587e17b91bcdf4e863d4ed9fc5)

This PR contains the following changes:

- Move previous implementation to a sub `v1/elitec` folder
- Add support for adapter v2
- Add support for Charybdis 3x6 format
- Add support for Dilemma
- Add support for RP2040-based MCU across all Bastard Keyboards
- Move keyboard USB IDs and strings to data driven
- Update headers to keep maintainers list up to date
- Run `qmk format-c`
- Minor no-op cleanups

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

0xcharly commented:
A couple updates:

- Split the PR into 2 commits to make it easier to review: moved the changes from `qmk format-c` into a separate commit
- Rebased onto latest `qmk:master`

drashna commented on keyboards/bastardkb/charybdis/3x5/config.h:
```diff
@@ -18,55 +18,25 @@
-
-/* serial.c configuration (for split keyboard) */
-#define SOFT_SERIAL_PIN D2
+/* Disable action features. */
+#define NO_ACTION_MACRO    // Disable old-style macro handling.
+#define NO_ACTION_FUNCTION // Disable old-style function handling.
```
These no longer exist, and can be completely removed. 
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/4x6/config.h:
```diff
@@ -18,52 +18,25 @@
-#define BOOTMAGIC_LITE_COLUMN 0
-#define BOOTMAGIC_LITE_ROW_RIGHT 5
-#define BOOTMAGIC_LITE_COLUMN_RIGHT 0
+/* Disable action features. */
+#define NO_ACTION_MACRO    // Disable old-style macro handling.
+#define NO_ACTION_FUNCTION // Disable old-style function handling.
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/skeletyl/config.h:
```diff
@@ -16,87 +17,34 @@
-#define RGBLED_NUM 36
-#define RGBLED_SPLIT { 18, 18 }
-#define RGBLIGHT_ANIMATIONS
+/* Disable action features. */
+#define NO_ACTION_MACRO    // Disable old-style macro handling.
+#define NO_ACTION_FUNCTION // Disable old-style function handling.
```
```suggestion
```

0xcharly replied:
Done.

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/tbkmini/config.h:
```diff
@@ -16,90 +17,34 @@
-#define RGBLED_NUM 42
-#define RGBLED_SPLIT { 21, 21 }
-#define RGBLIGHT_ANIMATIONS
+/* Disable action features. */
+#define NO_ACTION_MACRO    // Disable old-style macro handling.
+#define NO_ACTION_FUNCTION // Disable old-style function handling.
```
```suggestion
```

drashna reviewed (changes requested):
Also, rgb matrix modes removed, without being re-added elsewhere?

0xcharly commented:
> Also, rgb matrix modes removed, without being re-added elsewhere?

Correct, they were added back in a follow-up planned PR.  Fixed and added back in this one.

(apologies for the delayed answer.)

drashna commented:
bump?

drashna commented on keyboards/bastardkb/charybdis/3x5/blackpill/rules.mk:
```diff
@@ -0,0 +1,42 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x5/keymaps/bstiq/config.h:
```diff
@@ -0,0 +1,184 @@
+#ifdef VIA_ENABLE
+/* Via/Vial configuration. */
+#    define DYNAMIC_KEYMAP_LAYER_COUNT 8
+#    ifdef VIAL_ENABLE
+/** Vial-specific configuration. */
+
+#        define VIAL_KEYBOARD_UID \
+            { 0x08, 0x7F, 0xDE, 0x7F, 0xFA, 0x71, 0xB7, 0x49 }
+#        define VIAL_UNLOCK_COMBO_ROWS \
+            { 0, 4 }
+#        define VIAL_UNLOCK_COMBO_COLS \
+            { 0, 0 }
+
+#        ifndef __arm__
+/** Disable unused vial features. */
+
+// In addition to RGB Matrix effects, VialRGB also provides direct LED control
+// with a script running on your computer.  Remove to reenable.
+#            define VIALRGB_NO_DIRECT
+#        endif // __arm__
+#    endif     // VIAL_ENABLE
```
Vial config needs to be removed.
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x5/keymaps/bstiq/vial.json:
```diff
@@ -0,0 +1,76 @@
+{
```
vial config needs to be removed

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x5/v2/elitec/rules.mk:
```diff
@@ -0,0 +1,38 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x6/blackpill/rules.mk:
```diff
@@ -0,0 +1,40 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x6/v1/elitec/rules.mk:
```diff
@@ -0,0 +1,36 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x6/v2/elitec/rules.mk:
```diff
@@ -0,0 +1,36 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/4x6/blackpill/rules.mk:
```diff
@@ -0,0 +1,41 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/4x6/v2/elitec/rules.mk:
```diff
@@ -0,0 +1,37 @@
+
+POINTING_DEVICE_ENABLE = yes # Enable trackball
+POINTING_DEVICE_DRIVER = pmw3360
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/dilemma/keymaps/bstiq/config.h:
```diff
@@ -0,0 +1,163 @@
+#ifdef VIA_ENABLE
+/* Via/Vial configuration. */
+#    define DYNAMIC_KEYMAP_LAYER_COUNT 8
+#    ifdef VIAL_ENABLE
+/** Vial-specific configuration. */
+
+#        define VIAL_KEYBOARD_UID \
+            { 0x09, 0x80, 0xDF, 0x80, 0xFB, 0x72, 0xB8, 0x4A }
+#        define VIAL_UNLOCK_COMBO_ROWS \
+            { 0, 4 }
+#        define VIAL_UNLOCK_COMBO_COLS \
+            { 0, 0 }
+#    endif // VIAL_ENABLE
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/scylla/blackpill/rules.mk:
```diff
@@ -0,0 +1,37 @@
+
+SPLIT_KEYBOARD = yes
+
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/skeletyl/blackpill/rules.mk:
```diff
@@ -0,0 +1,38 @@
+SPLIT_KEYBOARD = yes
+LAYOUTS = split_3x5_3
+
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/tbkmini/blackpill/rules.mk:
```diff
@@ -0,0 +1,38 @@
+SPLIT_KEYBOARD = yes
+LAYOUTS = split_3x6_3
+
+# https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x5/v1/elitec/rules.mk:
```diff
@@ -36,3 +32,7 @@ POINTING_DEVICE_ENABLE = yes # Enable trackball
 POINTING_DEVICE_DRIVER = pmw3360
 # https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/4x6/v1/elitec/rules.mk:
```diff
@@ -35,3 +31,7 @@ POINTING_DEVICE_ENABLE = yes # Enable trackball
 POINTING_DEVICE_DRIVER = pmw3360
 # https://qmk.fm/changes/2018-11-16-use-a-single-endpoint-for-hid-reports
```
```suggestion
```

0xcharly replied:
Done.

drashna commented on keyboards/bastardkb/charybdis/3x5/keymaps/bstiq/rules.mk:
```diff
@@ -0,0 +1,12 @@
+endif
+
+VIA_ENABLE = yes
+VIAL_ENABLE = yes
+VIALRGB_ENABLE = yes
```
```suggestion
```

0xcharly replied:
Done.

0xcharly commented:
A few notes that summaries the various conversations I had with @drashna, @filterpaper, and @karlk90.

## Stability
At this stage, those firmwares are rather well tested and considered stable. They are used daily by multiple members of our community, and are shipped with @bstiq's prebuilds (bastardkb.com).

They have been available through the github.com/bastardkb/bastardkb-qmk fork for close to a year, and have evolved to adopt the latest and greatest from QMK.

## Freshness
To avoid the firmwares becoming stale, our fork is constantly updated against the latest changes that are upstreamed on qmk:master and qmk:develop, and against the latest designs and improvements that @drashna and @karlk90 always kindly notifies us of.

We have been amongst the first few to add support for the STeMCell, and more recently the RP2040.  We have integrated the recent improvements to the pointing device driver, and the latest feature for the Circque trackpad.

## Converters
We've had several discussions with @drashna and @filterpaper w.r.t folder structure and use of converters instead of maintaining several sub folders for each boards.

- While the converters route seems promising, it does not seem to be fully adequate _yet_ for BastardKB boards: we support multiple boards (pro micro, blackpill, elite-c, Splinky, STeMCell) for each keyboards, and while some of these share the same physical footprint (Elite-c/Splinky/STeMCell), these boards do not share the same feature set (mostly because we have to cut down a lot of features on AVR), nor do they always share the same configuration (_eg._ WS2812 driver configuration for the STeMCell). Using a multi folder structure simplifies the maintenance, and the build documentation
- Current converters seem to target the pro micro footprint, which does not include the bottom 5 pins that are available on boards that share the elite-c footprint, which the Bastard Keyboard line use, so converters as they stand today would only partially factorize the configuration, while still requiring per-board setup
- While the current structure and configurations are now well tested and I have a high level of confidence that they work as expected, I would not be able to say the same if we were to make changes now to these keyboard configurations (there are 26 different configurations, and multiple keymaps that would need retesting after such changes)

That being said, I am looking forward to the progress the core team is making with converters, and to being able to take advantage of them for the Bastard Keyboard firmwares. However, I believe this will be best addressed in future PR, to avoid further delaying the upstream of the Bastard Keyboard firmware implementations.

0xcharly commented:
Thanks for the review @drashna, let me cleanup the Vial-related stuff.

keyboard-magpie reviewed (changes requested):
Very impressive work overall, my only request is that the readme level of detail (keyboards/bastardkb/charybdis/3x6/keymaps/via/readme.md) for charybdis 3x6 is replicated  for the 3x5 and 4x6 via keymaps as they are complex keymaps with key features that should be highlighted to the end user, it should be as simple as duplicating the readme for the 3x6 via keymap.

bstiq commented:
Excellent work @0xcharly , thanks !
Looks good to me.
